src/ppx_variants_conv.ml

```diff
@@ -4,619 +4,830 @@
+         let fresh_var = supply_excluding bound in
+         List.map all ~f:(fun core_ty ->
+           match core_ty.ptyp_desc with
+           | Ptyp_any -> {core_ty with ptyp_desc = Ptyp_var (fresh_var ()) }
+           | _ -> core_ty
+         )
+       in
+       ptyp_constr ~loc {loc;txt= Longident.Lident td.ptype_name.txt } params 
+     ;;
+
+     (** Replace tyvars with type constructors corresponding to newtypes 
+         declaration for locally abstract types *)
+     let newtypes ty =
+       let rec aux acc ty =
+         let acc, ptyp_desc =
+         match ty.ptyp_desc with
+         | Ptyp_any -> acc , ty.ptyp_desc
+         | Ptyp_var nm -> 
+           let lblloc = {loc = ty.ptyp_loc; txt = nm} in
+           let ty = Ptyp_constr ({loc = ty.ptyp_loc; txt = Lident nm}, []) in
+           lblloc::acc , ty
+         | Ptyp_alias (inner_ty, lbl) -> 
+           let acc , inner_ty = aux acc inner_ty in
+           acc , Ptyp_alias (inner_ty, lbl)
+         | Ptyp_arrow (arg_lbl, ty1, ty2) ->
+           let acc , ty1 = aux acc ty1 in
+           let acc , ty2 = aux acc ty2 in 
+           acc, Ptyp_arrow (arg_lbl, ty1, ty2)
+         | Ptyp_tuple tys -> 
+           let acc , tys = List.fold_map tys ~init:acc ~f:aux in
+           acc, Ptyp_tuple tys 
+         | Ptyp_constr (nm, tys) -> 
+           let acc , tys = List.fold_map tys ~init:acc ~f:aux in 
+           acc,  Ptyp_constr (nm, tys)
+         | Ptyp_class (nm, tys) -> 
+           let acc , tys = List.fold_map tys ~init:acc ~f:aux in 
+           acc, Ptyp_class (nm, tys)
+         | Ptyp_object (flds, closed_flag) ->
+           let acc, flds = List.fold_map flds ~init:acc ~f:aux_obj_fld in 
+           acc, Ptyp_object (flds, closed_flag)
+         | Ptyp_variant (row_flds, closed_flag, labels_opt) ->
+           let acc, row_flds = List.fold_map row_flds ~init:acc ~f:aux_row_fld in 
+           acc, Ptyp_variant (row_flds, closed_flag, labels_opt)
+         | Ptyp_poly (vars, ty) -> 
+           let acc, ty = aux acc ty in
+           acc, Ptyp_poly (vars, ty)
+         | Ptyp_package _ | Ptyp_extension _ ->
+           raise_unsupported_ty ty.ptyp_loc
+       in
+       acc, {ty with ptyp_desc} 
+     and aux_obj_fld acc fld =
+       let acc, pof_desc =
+         match fld.pof_desc with
+         | Otag (lbl, ty) -> 
+           let acc, ty = aux acc ty in 
+           acc, Otag (lbl, ty)
+         | Oinherit ty -> 
+           let acc, ty = aux acc ty in 
+           acc, Oinherit ty
+       in
+       acc, {fld with pof_desc}
+     and aux_row_fld acc fld =
+       let acc, prf_desc=
+         match fld.prf_desc with
+         | Rtag (lbl, flag, tys) -> 
+           let acc, tys = List.fold_map tys ~init:acc ~f:aux in
+           acc, Rtag (lbl, flag, tys)
+         | Rinherit ty -> 
+           let acc, ty = aux acc ty in
+           acc, Rinherit ty
+       in
+       acc, {fld with prf_desc}
+     in
+     aux [] ty
+   ;;
```

Comment:
Isn't it possible to write this in 3 lines by inheriting from `Ast_traverse.fold` or similar?